test/apps/iodemo/ucx_wrapper.cc

```diff
@@ -0,0 +1,591 @@
+        _err_str += func_name + "() failed: " + ucs_status_string(status) +
+                    "\n";
+    }
+
+    UcxError(const UcxConnection &conn, ucs_status_t status) :
+        _status(status) {
+        std::stringstream ss;
+        ss << "Connection " << &conn << " is closed with status "
+           << ucs_status_string(_status) << std::endl;
+        _err_str += ss.str();
+    }
+
+    ~UcxError() _GLIBCXX_NOTHROW {
+    }
+
+    const char* what() const _GLIBCXX_NOTHROW {
+        return _err_str.c_str();
+    }
+
+    const char* status() const _GLIBCXX_NOTHROW {
+        return ucs_status_string(_status);
+    }
+private:
+    std::string  _err_str;
+    ucs_status_t _status;
+};
+
+void UcxCallback::operator()()
+{
+}
+
+UcxContext::UcxContext(size_t iomsg_size) :
+    _listener(NULL), _iomsg_buffer(iomsg_size, '\0')
+{
+    /* Create context */
+    ucp_params_t ucp_params;
+    ucp_params.field_mask   = UCP_PARAM_FIELD_FEATURES |
+                              UCP_PARAM_FIELD_REQUEST_INIT |
+                              UCP_PARAM_FIELD_REQUEST_SIZE;
+    ucp_params.features     = UCP_FEATURE_TAG |
+                              UCP_FEATURE_STREAM;
+    ucp_params.request_init = request_init;
+    ucp_params.request_size = sizeof(ucx_request);
+    ucs_status_t status = ucp_init(&ucp_params, NULL, &_context);
+    if (status != UCS_OK) {
+        throw UcxError("ucp_init", status);
+    }
+
+    log() << "created context " << _context << std::endl;
+
+    /* Create worker */
+    ucp_worker_params_t worker_params;
+    worker_params.field_mask  = UCP_WORKER_PARAM_FIELD_THREAD_MODE;
+    worker_params.thread_mode = UCS_THREAD_MODE_SINGLE;
+    status = ucp_worker_create(_context, &worker_params, &_worker);
+    if (status != UCS_OK) {
+        ucp_cleanup(_context);
+        throw UcxError("ucp_worker_create", status);
+    }
+
+    log() << "created worker " << _worker << std::endl;
+
+    post_recv();
+}
+
+UcxContext::~UcxContext()
+{
+    while (!_closing_conns.empty()) {
+        UcxConnection *conn = _closing_conns.front();
+        _closing_conns.pop_front();
+        try {
+            delete conn;
+        } catch (UcxError &e) {
+            debug() << e.what();
+        }
+    }
+
+    while (!_conn_requests.empty()) {
+        assert(_listener);
+        debug() << "reject connection request " << _conn_requests.front();
+        ucp_listener_reject(_listener, _conn_requests.front());
+        _conn_requests.pop_front();
+    }
+
+    // since all connections are handled, close the listener before next
+    // progress to avoid new incoming requests
+    if (_listener) {
+        ucp_listener_destroy(_listener);
+    }
+
+    ucp_request_cancel(_worker, _iomsg_recv_request);
+    try {
+        request_wait("iomsg_recv", _iomsg_recv_request);
+    } catch (const UcxError& e) {
+        debug() << "io msg request is completed with status " << e.status()
+                << std::endl;
+    }
```

Comment:
maybe add private helper function for every object that we close: close_connections, reject_conn_requests, cancel_recvs, etc.

Reply:
to use only once? maybe add class wrappers for such objects with own ctor/dtor?

Reply:
to use only once, will be simpler and self-explanatory

Reply:
`class UcxContext` is over complicated and supposed for inheritance in app.
such methods must be called in certain order and only on destruction (at least for now).
adding it could push users to use it just by name, change `private` to `public` or add `friend` word much easier than debug side effects in corner cases later.